syntax = "proto3";

package newsletter;

option java_multiple_files = true;

// ─────────────────────────────────────────────
// Service
// ─────────────────────────────────────────────

// Newsletter email rendering and delivery service.
service NewsletterService {
  // Render the newsletter email and send it to the recipient.
  rpc SendNewsletter (SendNewsletterRequest) returns (SendNewsletterResponse);
}

// ─────────────────────────────────────────────
// Request / Response
// ─────────────────────────────────────────────

message SendNewsletterRequest {
  // Recipient
  string recipient_email = 1;
  string recipient_name  = 2;

  // Header
  string date           = 3; // e.g. "2026年2月8日 · 星期日"
  int32  edition_number = 4;

  // Sections
  WeatherForecast          weather         = 5;
  repeated NewsItem        top_news        = 6;
  repeated StockInfo       stocks          = 7;
  repeated ContentSection  custom_sections = 9; // extensible
  repeated HNStory         hn_stories      = 10;
}

message SendNewsletterResponse {
  bool   success    = 1;
  string message_id = 2; // from the email provider
  string error      = 3; // populated on failure
}

// ─────────────────────────────────────────────
// Domain messages
// ─────────────────────────────────────────────

message WeatherForecast {
  string location     = 1;
  string condition    = 2;
  string icon         = 3;
  int32  temp_current = 4;
  int32  temp_high    = 5;
  int32  temp_low     = 6;
  string summary      = 7;
  // Astronomy (merged)
  string sunrise      = 8;
  string sunset       = 9;
  string day_length   = 10;
  string golden_hour  = 11;
  string astro_note   = 12;
  // 3-day forecast
  repeated ForecastDay forecasts = 13;
}

message ForecastDay {
  string label     = 1; // e.g. "周一"
  string icon      = 2; // emoji
  string condition = 3;
  int32  high      = 4;
  int32  low       = 5;
}

message NewsItem {
  string headline = 1;
  string summary  = 2;
  string source   = 3; // e.g. "Reuters"
  string url      = 4;
  string category = 5; // e.g. "Technology", "World"
}

message StockInfo {
  string symbol       = 1; // e.g. "AAPL"
  string company_name = 2;
  double price        = 3;
  double change        = 4; // absolute $ change
  double change_percent = 5; // e.g. 1.23 means +1.23%
}

// Generic extensible section – any new content type can be
// represented as a ContentSection without changing the proto.
message ContentSection {
  string title        = 1;
  string content_html = 2; // rendered HTML snippet
  string icon         = 3; // optional emoji / icon
}

// Hacker News top story
message HNStory {
  string title         = 1;
  string url           = 2;
  int32  points        = 3;
  int32  comment_count = 4;
  string hn_url        = 5; // link to HN discussion
  string title_cn      = 6; // Chinese translation
}